lock-impl/src/test/java/com/palantir/lock/impl/LegacyTimelockServiceIntegrationTest.java

```diff
@@ -0,0 +1,156 @@
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public final class LegacyTimelockServiceIntegrationTest {
+
+    private static final LockClient LOCK_CLIENT = LockClient.of("foo");
+    private static final TimestampLeaseName TIMESTAMP_LEASE_NAME1 = TimestampLeaseName.of("lease1");
+    private static final TimestampLeaseName TIMESTAMP_LEASE_NAME2 = TimestampLeaseName.of("lease2");
+
+    private LockServiceImpl lockService;
+    private LegacyTimelockService timelock;
+
+    @BeforeEach
+    public void beforeEach() {
+        lockService = LockServiceImpl.create();
+        timelock = new LegacyTimelockService(new InMemoryTimestampService(), lockService, LOCK_CLIENT);
+    }
+
+    @AfterEach
+    public void afterEach() {
+        lockService.close();
+    }
+
+    @Test
+    public void canAcquireAndReleaseMultipleLeases() {
+        TestLease lease1 = acquire(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2);
+        TestLease lease2 = acquire(TIMESTAMP_LEASE_NAME1);
+        TestLease lease3 = acquire(TIMESTAMP_LEASE_NAME2);
+        TestLease lease4 = acquire(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2);
+        assertThat(lease1.minLeasedTimestamp())
+                .isEqualTo(lease2.minLeasedTimestamp())
+                .isEqualTo(lease3.minLeasedTimestamp())
+                .isEqualTo(lease4.minLeasedTimestamp());
+        assertThatMinLeaseTimestamps(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2)
+                .containsEntry(TIMESTAMP_LEASE_NAME1, lease1.minLeasedTimestamp())
+                .containsEntry(TIMESTAMP_LEASE_NAME2, lease1.minLeasedTimestamp());
+
+        // Release middle leases
+        lease2.unlock();
+        assertThatMinLeaseTimestamps(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2)
+                .containsEntry(TIMESTAMP_LEASE_NAME1, lease1.minLeasedTimestamp())
+                .containsEntry(TIMESTAMP_LEASE_NAME2, lease1.minLeasedTimestamp());
+        lease3.unlock();
+        assertThatMinLeaseTimestamps(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2)
+                .containsEntry(TIMESTAMP_LEASE_NAME1, lease1.minLeasedTimestamp())
+                .containsEntry(TIMESTAMP_LEASE_NAME2, lease1.minLeasedTimestamp());
+
+        // Release oldest lease
+        lease1.unlock();
+
+        // TODO(jakubk): This is awkward because we don't have the leased timestamps in the response.
+        MapAssert<TimestampLeaseName, Long> timestampsAssert =
+                assertThatMinLeaseTimestamps(TIMESTAMP_LEASE_NAME1, TIMESTAMP_LEASE_NAME2);
+        timestampsAssert.extractingByKey(TIMESTAMP_LEASE_NAME1).satisfies(minLeaseTimestamp -> assertThat(
+                        minLeaseTimestamp)
+                .isGreaterThan(lease4.minLeasedTimestamp()));
+        timestampsAssert.extractingByKey(TIMESTAMP_LEASE_NAME2).satisfies(minLeaseTimestamp -> assertThat(
+                        minLeaseTimestamp)
+                .isGreaterThan(lease4.minLeasedTimestamp()));
+    }
```

Comment:
Should we have some way to validate we use fresh timestamps when we grab leases?

If there is a degenerate impl and leasing is done with the same timestamp throughout, the test I suspect still passes.

This is awkward because of the same reason you cite.

But you can add say in the last asserts: .isStrictlyGreaterThan(leaseN.minLeasedTimestamp()) for N < 4

And in other places if you so wish

Reply:
> Should we have some way to validate we use fresh timestamps when we grab leases?

I don't want to check for that, that feels like too much. I don't think tests should be necessarily adversarial to the implementation: like what else would you use if not fresh timestamps :) That's basically the whole point of timestamp service :)